File: 303-RangeSumQuery-Immutable/303-RangeSumQuery-Immutable.py
# Last updated: 8/1/2026, 12:56:31 PM
1class Solution(object):
2    def isPerfectSquare(self, num):
3        if num < 2:
4            return True
5
6        left, right = 2, num // 2
7        while left <= right:
8            mid = (left + right) // 2
9            squared = mid * mid
10            if squared == num:
11                return True
12            elif squared < num:
13                left = mid + 1
14            else:
15                right = mid - 1
16
17        return False

File: 303-RangeSumQuery-Immutable/303-RangeSumQuery-Immutable.java
// Last updated: 8/1/2026, 1:01:44 PM
1class Solution {
2    public int sumOfLeftLeaves(TreeNode root) {
3        if (root == null) {
4            return 0;
5        }
6        
7        Queue<Pair<TreeNode, Boolean>> queue = new LinkedList<>();
8        queue.offer(new Pair<>(root, false));  // (node, is_left)
9        int totalSum = 0;
10        
11        while (!queue.isEmpty()) {
12            Pair<TreeNode, Boolean> pair = queue.poll();
13            TreeNode node = pair.getKey();
14            boolean isLeft = pair.getValue();
15            
16            if (isLeft && node.left == null && node.right == null) {
17                totalSum += node.val;
18            }
19            
20            if (node.left != null) {
21                queue.offer(new Pair<>(node.left, true));
22            }
23            if (node.right != null) {
24                queue.offer(new Pair<>(node.right, false));
25            }
26        }
27        
28        return totalSum;
29    }
30}